### DISCUSSION TITLE
z3 performance issue

### DISCUSSION THREAD

**Discussion by usagitoneko97** [Z3Prover/z3#5057] [Q&A] [answered]
For smt2-lib file: [stuck2.smt.txt](https://github.com/Z3Prover/z3/files/6049267/stuck2.smt.txt) , I find that it's very slow (1.5 s) on z3, but it run instantaneous on cvc4. Will setting theory or strategy in z3 helps? If so, what should I set?

In the constraints generated and dumped, i see a lot of intermediate variable being created. Could that be the cause?
```
 (let (($x624 (>= 4 CM)))
 (let (($x506 (not $x624)))
 (let ((?x844 (div ?x756 128)))
 (let ((?x897 (+ ?x844 1)))
 (let (($x687 (<= 4 ?x897)))
 (let (($x344 (= 0 (mod ?x536 128))))
```

Or in general looking at the smt2-lib file, is there anything I can do to speed up z3 performance?

**Comment by NikolajBjorner** [comment, marked as answer]
they are let-bound variables, not intermediate variables
set the option smt.arith.solver=2, or in the file (set-option :smt.arith.solver 2) and it is much faster.
This is an instance where the new arithmetical solver suffers from a performance overhead not shared with the legacy arithmetic solver.
